File: bound/kt/src/test/kotlin/web5/sdk/crypto/keys/JwkTest.kt
package web5.sdk.crypto.keys

import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.fail
import web5.sdk.Web5Exception

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class JwkTest {

    @Test
    fun test_ec_valid() {
        val jwk = Jwk(
            kty = "EC",
            crv = "secp256k1",
            x = "x_value",
            y = "y_value"
        )

        val thumbprint = jwk.computeThumbprint()
        assertEquals("yiiszVT5Lwt6760MW19cHaJ61qJKIfe20sUW8dNxBv4", thumbprint)
    }

    @Test
    fun test_okp_valid() {
        val jwk = Jwk(
            kty = "OKP",
            crv = "Ed25519",
            x = "x_value"
        )

        val thumbprint = jwk.computeThumbprint()
        assertEquals("nDMRVZm4lpedGjuJGO4y3YVJJ0krDF0aSz4KhlncDdI", thumbprint)
    }

    @Test
    fun test_unsupported_kty() {
        val jwk = Jwk(
            kty = "RSA",
            crv = "RS256",
            x = "x_value",
            y = "y_value"
        )

        val exception = assertThrows<Web5Exception> {
            jwk.computeThumbprint()
        }

        assertEquals("data member error kty not supported RSA", exception.message)
    }

    @Test
    fun test_empty_kty() {
        val jwk = Jwk(
            kty = "",
            crv = "Ed25519",
            x = "x_value"
        )

        val exception = assertThrows<Web5Exception> {
            jwk.computeThumbprint()
        }

        assertEquals("data member error kty cannot be empty", exception.message)
    }

    @Test
    fun test_empty_x() {
        val jwk = Jwk(
            kty = "OKP",
            crv = "Ed25519",
            x = ""
        )

        val exception = assertThrows<Web5Exception> {
            jwk.computeThumbprint()
        }

        assertEquals("data member error x cannot be empty", exception.message)
    }

    @Test
    fun test_empty_crv() {
        val jwk = Jwk(
            kty = "EC",
            crv = "",
            x = "x_value",
            y = "y_value"
        )

        val exception = assertThrows<Web5Exception> {
            jwk.computeThumbprint()
        }

        assertEquals("data member error crv cannot be empty", exception.message)
    }

    @Test
    fun test_ec_missing_y() {
        val jwk = Jwk(
            kty = "EC",
            crv = "P-256",
            x = "x_value"
        )

        val exception = assertThrows<Web5Exception> {
            jwk.computeThumbprint()
        }

        assertEquals("data member error missing y", exception.message)
    }

    @Test
    fun test_ec_empty_y() {
        val jwk = Jwk(
            kty = "EC",
            crv = "P-256",
            x = "x_value",
            y = ""
        )

        val exception = assertThrows<Web5Exception> {
            jwk.computeThumbprint()
        }

        assertEquals("data member error y cannot be empty", exception.message)
    }
}

File: bound/typescript/tests/crypto/jwk.test.ts
import { expect } from 'chai';
import { Jwk } from '../../src/crypto/jwk';
import { Web5Error } from '../../src/errors';

describe('Jwk class', () => {
  it('should compute the correct thumbprint for EC key', async () => {
    const jwk = new Jwk('ES256', 'EC', 'secp256k1', undefined, 'x_value', 'y_value');
    const thumbprint = jwk.computeThumbprint();
    expect(thumbprint).to.equal('yiiszVT5Lwt6760MW19cHaJ61qJKIfe20sUW8dNxBv4');
  });

  it('should compute the correct thumbprint for OKP key', async () => {
    const jwk = new Jwk(undefined, 'OKP', 'Ed25519', undefined, 'x_value', undefined);
    const thumbprint = jwk.computeThumbprint();
    expect(thumbprint).to.equal('nDMRVZm4lpedGjuJGO4y3YVJJ0krDF0aSz4KhlncDdI');
  });

  it('should throw error for unsupported kty', async () => {
    try {
      const jwk = new Jwk(undefined, 'RSA', 'RS256', undefined, 'x_value', 'y_value');
      jwk.computeThumbprint();
    } catch (error: any) {
      expect(error.variant).to.equal('DataMember');
      expect(error.message).to.equal('data member error kty not supported RSA');
    }
  });

  it('should throw error when kty is empty', async () => {
    try {
      const jwk = new Jwk(undefined, '', 'Ed25519', undefined, 'x_value', undefined);
      jwk.computeThumbprint();
    } catch (error: any) {
      expect(error.variant).to.equal('DataMember');
      expect(error.message).to.equal('data member error kty cannot be empty');
    }
  });

  it('should throw error when x is empty', async () => {
    try {
      const jwk = new Jwk(undefined, 'OKP', 'Ed25519', undefined, '', undefined);
      jwk.computeThumbprint();
    } catch (error: any) {
      expect(error.variant).to.equal('DataMember');
      expect(error.message).to.equal('data member error x cannot be empty');
    }
  });

  it('should throw error when crv is empty', async () => {
    try {
      const jwk = new Jwk(undefined, 'EC', '', undefined, 'x_value', 'y_value');
      jwk.computeThumbprint();
    } catch (error: any) {
      expect(error.variant).to.equal('DataMember');
      expect(error.message).to.equal('data member error crv cannot be empty');
    }
  });

  it('should throw error when y is missing for EC key', async () => {
    try {
      const jwk = new Jwk(undefined, 'EC', 'P-256', undefined, 'x_value', undefined);
      jwk.computeThumbprint();
    } catch (error: any) {
      expect(error.variant).to.equal('DataMember');
      expect(error.message).to.equal('data member error missing y');
    }
  });

  it('should throw error when y is empty for EC key', async () => {
    try {
      const jwk = new Jwk(undefined, 'EC', 'P-256', undefined, 'x_value', '');
      jwk.computeThumbprint();
    } catch (error: any) {
      expect(error.variant).to.equal('DataMember');
      expect(error.message).to.equal('data member error y cannot be empty');
    }
  });
});